### PR TITLE
[WIP]gnrc_tcp

### PR DESCRIPTION
**Pull request by brummer-simon**
Current State of ng_tcp. This Branch is currently under heavy development. Your comments on the current state are very welcome :). 

This PR is intendend for reference in the NSTF Todo List.

### CODE REVIEW

File: sys/include/net/ng_tcp.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+/**
+ * @brief Size of a TCP header
+ */
+// TODO: Depends on Headeroptions.
+#define NG_TCP_HDR_LENGTH   (20U)
+
+/**
+ *  @brief States for the TCP State Maschine
+ */
+typedef enum {CLOSED, LISTEN, SYN_SENT, SYN_RCVD, ESTABLISHED} ng_tcp_states_t;
```
Comment (OlegHahm) [inline sys/include/net/ng_tcp.h]: For readability and documentation I would prefer to put each enum into a single line.

File: sys/net/transport_layer/ng_tcp/ng_tcp.c [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    return 0;
+}
+
+int8_t ng_tcp_passive_open(ng_tcp_tcb_t *tcb, uint16_t local_port)
+{   
+    msg_t msg;
+    ng_tcp_hdr_t  tcp_hdr;
+    ng_pktsnip_t* tcp_snp = NULL;
+
+    // Guard: tcb must be in state CLOSED
```
Comment (OlegHahm) [inline sys/net/transport_layer/ng_tcp/ng_tcp.c]: Please don't use C++ comment style in C code.

**Comment by OlegHahm** [comment]
@cgundogan, maybe you are interested, too.

**Comment by authmillenon** [comment]
Please don't use sub-directories in `tests` (except unittests). See e.g. thread and vtimer tests.

File: sys/net/crosslayer/ng_netreg/ng_netreg.c [hunk 3]
```diff
@@ -22,6 +22,10 @@
 #include "net/ng_pkt.h"
 #include "net/ng_ipv6.h"
 
+#ifdef MODULE_NG_TCP
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_netreg/ng_netreg.c]: ifdefs should not be necessary here

**Comment by OlegHahm** [comment]
@brummer-simon, what's the latest on this one?

**Comment by brummer-simon** [comment]
Currently I am running in a bunch of problems with implementing TCP. I suspect the cause of it in RIOTs scheduling model and timer implementation. The non-premetive scheduling model seems to be the biggest obstacle because RFC 793 assumes a premetive scheduler. 

Currently I'll write it down and up the pdf up for discussion.

**Comment by OlegHahm** [comment]
> The non-premetive scheduling model seems to be the biggest obstacle because RFC 793 assumes a premetive scheduler. 

Huh? Since when does RIOT has a non-preemptive scheduler? Don't confuse the fact that RIOT's scheduler works without timeslices to interrupt threads periodically with the fact threads are preemptible.

**Comment by brummer-simon** [comment]
Okay. Still the Problem is, main Function runs as long as it doesn't put itself to sleep or wait for a message or what ever. After that TCP is able to do actual work. And this is a Problem. The TCP-Thread must be able react on incomming data as soon as possible, independently from userspace.

**Comment by kaspar030** [comment]
> Okay. Still the Problem is, main Function runs as long as it doesn't put itself to sleep or wait for a message or what ever. After that TCP is able to do actual work. And this is a Problem. The TCP-Thread must be able react on incomming data as soon as possible, independently from userspace.

Give TCP a higher priority (lower priority value) than main, if your
main does busy-waiting.

**Comment by brummer-simon** [comment]
Okay as general scheduling questing. If my TCP-Thread has a higher priority than my main-Thread and the TCP-Thread is currently waiting for a message, the main thread is running. 

If a message for the tcp thread commes in, will there be a context switch to the tcp thread as soon as the message comes in or after the main thread puts it self to sleep ?

**Comment by OlegHahm** [comment]
> If a message for the tcp thread commes in, will there be a context switch to the tcp thread as soon as the message comes in or after the main thread puts it self to sleep ?

As soon as the message comes in.

**Comment by brummer-simon** [comment]
Ah okay. Then I see why its not a non-premptive scheduler. Still i'll have to figure out where exactly my Problem is.

**Comment by OlegHahm** [comment]
Let us know as soon as you've figured out what the problem is. Maybe we can help.

**Comment by kaspar030** [comment]
> Still i'll have to figure out where exactly my Problem is.

Are you testing on native?

**Comment by brummer-simon** [comment]
Native and the samr-xpro board.

**Comment by kaspar030** [comment]
> Native and the samr-xpro board.

Are you using the shell? And if yes, are you using uart0?

**Comment by brummer-simon** [comment]
Yes and Yes. Currently I focus more on Bugfixing on native.

**Comment by gebart** [comment]
Also note that in order to give a thread a _higher_ priority means that you
assign it a _lower_ number for the priority value IIRC.

**Comment by brummer-simon** [comment]
Okay followup question about scheduling. 

My main thread is waiting for a message. the tcp thread is running and it is sending a message to the main thread. 

The priority of the tcp thread is higher than main(a lower number). After the message was sent from tcp to main, main takes over. That means the priority is raised after getting a message from a higher prioritzed thread. Is that right?

Ist there somehow a tutorial how the scheduler behaves?

**Comment by OlegHahm** [comment]
> The priority of the tcp thread is higher than main(a lower number). After the message was sent from tcp to main, main takes over. That means the priority is raised after getting a message from a higher prioritzed thread. Is that right?

Not quite. After TCP thread sent the message to main thread, TCP will keep running. Just the main thread will change its state to `PENDING`, i.e. it will continue to run as soon as no thread with a higher priority is in the run queue.

> Ist there somehow a tutorial how the scheduler behaves?

I'm not sure. At least there should be something in @kaspar030's thesis IIRC.

**Comment by brummer-simon** [comment]
Okay then I am really confused by my codes behavior.

**Comment by OlegHahm** [comment]
If you tell us, what and how you're testing, maybe I can try to shed some light on it.

**Comment by OlegHahm** [comment]
btw. I guess you have seen http://doc.riot-os.org/group__core__sched.html#details, right?

**Comment by brummer-simon** [comment]
Okay. I'll merge the latest changes into my branch, push it and tell you whats my problem is.

**Comment by kaspar030** [comment]
@brummer-simon The scheduling policy is super-simple: the thread with the highest priority (lowest value) will always be run, possibly preempting lower priority threads.

So if tcp has a higher priority, then sends a msg to main (which has a lower priority), main will be set to PENDING, but if tcp continues to do stuff, main will not get scheduled.

**Comment by kaspar030** [comment]
If you're using the shell, try to not use uart0. just don't "usemodule" uart0 and maybe merge  #3402.

**Comment by brummer-simon** [comment]
Okay. After the merge with current master I get segfaults in my retransmission mechanism. I'll take care of that, after that explain that I am testing right now.

**Comment by Kijewski** [comment]
Assuming there are threads L1, L2, L3 with the same low priority, and a thread H with a high priority, which is repeatedly message receive blocked.

* L1 is running and sends H a message. L1 gets preempted in favor of H.
* H is running for a bit and then waits for a message again.
* As there are no (runnable) threads with `[ priority(H) ; priority(L1) )`, L1 runs again.
* I.e. L2 and L3 may stave if L1 does not voluntarily yield. (thread_yield, msg_wait, msg_send [if the receivers queue is full], etc.).

**Comment by DipSwitch** [comment]
The segfault has to do with the the `nomac` driver releases the packet back to the packet buffer if it has send it. Then new data is wrote over the previous `gnrc_pktsnip_t` however the TCP tcb uses that the now overwritten pointers `cur_pkt` and `cur_tcp_hdr` to try and resend the data.

The SYN/ACK still is invalid though, the TCP header is correct but although `gncr_ipv6.c` sets the src address. When the packet is send however the src address, next header and payload are invalid / not set.

Any ideas?

**Comment by DipSwitch** [comment]
Related to issue: #3707

**Comment by brummer-simon** [comment]
Okay I pulled the current Master and DipSwitchs Branch that did the whole renaming(Thanks by the Way:) ) and squashed the hole thing. 

I'll try to get the handshake going

**Comment by DipSwitch** [comment]
The handshake is working :) I think we can start sending data :D

**Comment by OlegHahm** [comment]
Nice. :sunglasses:

**Comment by brummer-simon** [comment]
My original goal was to implement the connection termination first, to make sure the fsm work as it should be. But as i wrote yesterday, I focus the next months on my Bachelor Thesis.

**Comment by brummer-simon** [comment]
Oh the merge incorperated a lot files that aren't involved in the hole tcp thing. I'll fix that in the afternoon.

**Comment by DipSwitch** [comment]
Both are even important :) good luck with your thesis!

**Comment by authmillenon** [comment]
@DipSwitch please comment in the PR not the commit. Comments in commits get easily lost.

**Comment by authmillenon** [comment]
You probably need to rebase to current master. There are a lot of changes in here now, that are actually a part of master already.

**Comment by DipSwitch** [comment]
I rebased and forced updated, sorry not the best way possible...

**Comment by brummer-simon** [comment]
Yes I'll implement the close operation. Over the next weeks.

File: sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+/**
+ * @brief Save the TCP's threads PID for later reference
+ */
+static kernel_pid_t _tcp_pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief Head of tcb linked list. The eventloop needs 
+ *        to look up the affected tcbs in case of an incomming paket.
+ *        ng_tcp_tcb_init() adds tcb to this list.
```
Comment (cgundogan) [inline sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c]: s/ng/gnrc/

**Comment by OlegHahm** [comment]
The last time I tried (about 1 month ago), I was not able to get the test application to send anything (not even the handshake packets).
As far as I can see, what would be the "eventloop" (as it is called in other gnrc modules) corresponds to the _fsm function here. However, this seems not to implement a common event loop for all TCP connections, but just for one connection. This would require every TCP connection to run its own thread which is most likely a bad idea. I would advice to use a central eventloop (just similar to what other gnrc modules do) and make all calls to TCP based on gnrc netapi calls.

Some more "formal" comments:
* Defining such a long function as the current `_fsm()` static inside a header is IMO not a good style. Please move this to a C file.
* The code could need some [uncrustifying](https://github.com/RIOT-OS/RIOT/pull/4063)

**Comment by brummer-simon** [comment]
Hi Oleg, 

currently i'am doing buzy spliting the static includes into thier own c-Files, and doing some renaming. For your concern with the _fsm corresponance. 

TCPs statemaschine must be able to do translations, without the users interaction, in case a FIN-Packet was received. Moving from ESTABLISHED to CLOSE_WAIT. The Eventloop interacts only in case of Packet reception with the FSM. The Transmission control blocks are organized as linked list. If the eventloop receives a packet, it looks for the receivers pid. The pid part of an tcb as well. The eventloop searches for the receivers tcb and calls the fsm-function with this specific tcb. 

Long story short, only one eventloop is needed for multiple connections.

**Comment by brummer-simon** [comment]
Merge fuckup lead to this pull request: #4744

**Comment by cgundogan** [comment]
closed in favor of #4744
